src/main/transaction.c

```diff
@@ -1396,61 +1358,225 @@ LOCAL_HELPER void
 do_async_workload(tdata_t* tdata, cdata_t* cdata, thr_coord_t* coord,
 		stage_t* stage)
 {
-	struct async_data_s* adatas;
 	uint32_t t_idx = tdata->t_idx;
-	uint64_t n_adatas;
-	queue_t adata_q;
 
-	// thread 0 is designated to handle async calls, the rest can immediately
-	// terminate
+	// Thread 0 is designated to handle async calls, the rest can immediately
+	// terminate.
 	if (t_idx != 0) {
 		thr_coordinator_complete(coord);
 		return;
 	}
 
-	n_adatas = cdata->async_max_commands;
-	adatas =
-		(struct async_data_s*) cf_malloc(n_adatas * sizeof(struct async_data_s));
+	async_workload_init_coord(coord, stage);
+
+	uint32_t n_workers = cdata->async_max_commands;
+	async_data_t* adatas =
+		(async_data_t*)cf_calloc(sizeof(async_data_t), n_workers);
+	async_data_t* worker_datas[n_workers];
+
+	for (uint32_t i = 0; i < n_workers; i++) {
+		worker_datas[i] = &adatas[i];
+	}
+
+	uint64_t tps = stage->tps;
+
+	// Microseconds per txn.
+	uint64_t min_usleep = (double)1.0 / (double)tps * 1000000;
 
-	queue_init(&adata_q, n_adatas);
-	for (uint32_t i = 0; i < n_adatas; i++) {
-		struct async_data_s* adata = &adatas[i];
+	min_usleep = min_usleep > 0 ? 1 : min_usleep;
```

Comment:
This can only be 1 or 0, is that right? Why bother calculating it?

Reply:
Supposed to be the number of microseconds per transaction.